support for zone transfer for kubernetes

bradbeam opened the pull request:
Building on the work in #1191

```
[22:22:56] $ go test -v -run TestKubernetesXFR
=== RUN   TestKubernetesXFR
--- PASS: TestKubernetesXFR (0.00s)
	xfr_test.go:45: &{ResponseWriter:0x2d398f0 Rcode:0 Len:365 Msg:;; opcode: QUERY, status: NOERROR, id: 65471
		;; flags: qr aa ra; QUERY: 1, ANSWER: 9, AUTHORITY: 0, ADDITIONAL: 0
		
		;; QUESTION SECTION:
		;cluster.local.	IN	 AXFR
		
		;; ANSWER SECTION:
		cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 0 7200 1800 86400 30
		svc1.testns.svc.cluster.local.	5	IN	A	10.0.0.1
		svc6.testns.svc.cluster.local.	5	IN	AAAA	1234:abcd::1
		172-0-0-2.hdls1.testns.svc.cluster.local.	5	IN	A	172.0.0.2
		172-0-0-3.hdls1.testns.svc.cluster.local.	5	IN	A	172.0.0.3
		5678-abcd--1.hdls1.testns.svc.cluster.local.	5	IN	AAAA	5678:abcd::1
		5678-abcd--2.hdls1.testns.svc.cluster.local.	5	IN	AAAA	5678:abcd::2
		external.testns.svc.cluster.local.	5	IN	CNAME	ext.interwebs.test.
		cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 0 7200 1800 86400 30
		 Start:2017-11-27 22:23:15.177925535 -0600 CST m=+0.048178501}
PASS
ok  	github.com/coredns/coredns/plugin/kubernetes	0.060s
```

I think it could still use some polish and support for additional record types.

miekg commented on plugin/kubernetes/controller.go:
```diff
@@ -410,3 +415,18 @@ func (dns *dnsControl) GetNamespaceByName(name string) (*api.Namespace, error) {
 	return v1ns, nil
 }
+
+func (dns *dnsControl) Modified() int64 {
```
This doesn't actually tell you if it is modified? Should we just the `Serial(state request.Request) uint32` for this? Or change that interface?

miekg commented on plugin/kubernetes/handler.go:
```diff
@@ -30,29 +30,31 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 
-	switch state.Type() {
-	case "A":
+	switch state.QType() {
```
Type() should just return AXFR  - I hope? Makes the PR smaller because it cuts out of the churn here.

bradbeam replied:
I thought it made sense to switch to using the constants for comparison versus a string comparison.

miekg replied:
yes, probably faster as well - should do this in diff. PR

miekg commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -461,12 +468,3 @@ func match(a, b string) bool {
 	return s == "*" || s == "any"
 }
-
-const (
```
Please try (I know it is hard) to now do cleanups in the same PR

bradbeam replied:
Should I just defer these for a later PR then?

miekg replied:
yep

miekg commented on plugin/kubernetes/handler.go:
```diff
@@ -30,29 +30,31 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 			break
 		}
 		fallthrough
+	case dns.TypeAXFR:
```
In the *file* plugin there is this:
~~~
	if state.QType() == dns.TypeAXFR || state.QType() == dns.TypeIXFR {
		xfr := Xfr{z}
		return xfr.ServeDNS(ctx, w, r)
	}
~~~
I.e. we jump to another ServeDNS that handles all xfr logic

bradbeam replied:
Can you elaborate a bit?
What would the benefit be of implementing another ServeDNS to handle this?

miekg commented on plugin/kubernetes/setup.go:
```diff
@@ -201,6 +201,12 @@ func kubernetesParse(c *caddy.Controller) (*Kubernetes, dnsControlOpts, error) {
 				}
 				k8s.ttl = uint32(t)
+			case "transfer":
+				dests, err := TransferParse(c)
```
Can you not just call `file.TransferParse(c, false)`

bradbeam replied:
We could do that, but I didn't want to create a dependency between the file and Kubernetes plugin.

bradbeam replied:
Also, the `file.TransferParse` supports transfer from, which we don't want to support in the Kubernetes plugin.

bradbeam replied:
But I guess we could just ignore/err if we get froms returned. So that's an option.

miekg replied:
secondary also depends on file, if it becomes cyclic we move to pkg/...
transfer from is an error when called `secondary = false`

miekg commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,80 @@
-func (e *Kubernetes) Serial(state request.Request) uint32 {
-	return uint32(time.Now().Unix())
-}
+func (k *Kubernetes) Serial(state request.Request) uint32 { return uint32(k.APIConn.Modified()) }
```
ah you do it here. Ok, maybe `Modified()` does make sense; however the name 'modified` is somewhat misleading

johnbelamaric commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,80 @@
+	serviceList := k.APIConn.ServiceList()
+	for _, svc := range serviceList {
+		// Endpoint query or headless service
+		if svc.Spec.ClusterIP == api.ClusterIPNone {
```
You should check the service type not just the defined fields. For example https://github.com/coredns/coredns/pull/1246

bradbeam replied:
How do we want to handle when type is empty? 
In reading through api/v1/types.go and api/v1/defaults.go.SetDefaults_Service it seems like this shouldn't ever be the case where svc.Spec.Type is "", but there are some test cases that don't seem to have this defined.

bradbeam replied:
I've updated the tests to explicitly define the type.

bradbeam commented:
Added in some additional bits to include SRV records
```
		;; QUESTION SECTION:
		;cluster.local.	IN	 AXFR
		
		;; ANSWER SECTION:
		cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1511935515 7200 1800 86400 30
		svc1.testns.svc.cluster.local.	5	IN	A	10.0.0.1
		_http._tcp.svc1.testns.svc.cluster.local.	5	IN	SRV	0 100 80 svc1.testns.svc.cluster.local.
		svc6.testns.svc.cluster.local.	5	IN	AAAA	1234:abcd::1
		_http._tcp.svc6.testns.svc.cluster.local.	5	IN	SRV	0 100 80 svc6.testns.svc.cluster.local.
		172-0-0-2.hdls1.testns.svc.cluster.local.	5	IN	A	172.0.0.2
		_http._tcp.hdls1.testns.svc.cluster.local.	5	IN	SRV	0 25 80 172-0-0-2.hdls1.testns.svc.cluster.local.
		_http._tcp.hdls1.testns.svc.cluster.local.	5	IN	SRV	0 25 80 172-0-0-3.hdls1.testns.svc.cluster.local.
		_http._tcp.hdls1.testns.svc.cluster.local.	5	IN	SRV	0 25 80 5678-abcd--1.hdls1.testns.svc.cluster.local.
		_http._tcp.hdls1.testns.svc.cluster.local.	5	IN	SRV	0 25 80 5678-abcd--2.hdls1.testns.svc.cluster.local.
		172-0-0-3.hdls1.testns.svc.cluster.local.	5	IN	A	172.0.0.3
		5678-abcd--1.hdls1.testns.svc.cluster.local.	5	IN	AAAA	5678:abcd::1
		5678-abcd--2.hdls1.testns.svc.cluster.local.	5	IN	AAAA	5678:abcd::2
		external.testns.svc.cluster.local.	5	IN	CNAME	ext.interwebs.test.
		cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1511935515 7200 1800 86400 30
```

It kind of feels like we're absuing the etcd.msg package for passing around a dump of the zone information since we're not actually storing any of the values in etcd for this use case. Also,  extracting the record type from the service is pretty limited. I'm curious if it's worth adding some additional fields to the msg.Service struct ( such as record type ) to allow it to be more explicit and perhaps updating the RR method to return the proper record for the associated service.

johnbelamaric commented:
FYI use this for `fallthrough` case: https://github.com/infobloxopen/themis/blob/850a2e0fdc10362d587e5e81541b7458b17843ff/contrib/coredns/policy/policy.go#L623

bradbeam commented:
The above kind of works. I'm able to get transfers to work some of the times, other times it seems like the messages are being received by the secondary out of order which is causing the client to keep the connection open for >2s which is causing the readdeadline timeout to get reached. It _feels_ like the issue is with https://github.com/coredns/coredns/pull/1259/files#diff-347a95a2d36b7b336cb55cc146bd16d6R81 when we start writing the message back. I put some debugging in file/secondary.go and I can see the records being received out of order sometimes.  May need some help in tracking that down.

miekg commented on plugin/etcd/msg/type.go:
```diff
@@ -18,6 +19,10 @@ func (s *Service) HostType() (what uint16, normalized net.IP) {
 
 	ip := net.ParseIP(s.Host)
 
+	if s.Key != "" && strings.HasPrefix(Domain(s.Key), "_") {
```
Why do we need this?
HostType() can't return TypeSRV as descibed in the function docs.

bradbeam replied:
This is used in the backend_lookup.go AXFR() so we can transfer the SRV records. Otherwise, we're limited to transferring only the types that HostType() supports ( A/AAAA/CNAME ). I can update the comment to show that it's supported.

miekg commented on plugin/kubernetes/handler.go:
```diff
@@ -53,6 +57,48 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
+
+			fakeMsg := fakew.ReadMsg()
+			// Strip out SOA records
+			records = append(records, fakeMsg.Answer[1:len(fakeMsg.Answer)-1]...)
```
I don't understand what's happening here... Why the `if k.Fallthrough`? and why the `fakewriter`

Maybe it neater rename plugin.Xfr to plugin.All() or have a private function that just grabs all records.

johnbelamaric replied:
If fall through is set, then the k8s plugin does not know all of the records that are in the zone. Some other later plugin may be handling some records. So, we push a request down the plugin chain to gather up all those records that may be managed by someone else.

bradbeam replied:
Yeah, what John said.
Our use case is to have host records at `cluster.local`, but then k8s services at `xxx.namespace.svc.cluster.local`. So we use kubernetes and then a fallthrough to file based for host entries. 
By using the `fakewriter` we can let the next plugin write out results locally and then inject the data into the payload we send back to the client.

miekg replied:
hmm, I don't like the complexity this introduces. I thought it would be enough to have it just return the records it has in the cluster - not grabbing things from all over the place?

johnbelamaric replied:
Well, if we are doing a zone transfer shouldn't we include all the records in that zone, regardless of which plugin handles them? From the point of view of an external server, it doesn't care how we construct the records internally, it just needs to know that it has everything for the requested zone.

I don't think this is really any more complex than fall through in general - and allows help debugging fall through results.

miekg replied:
No, we should do this Fallthrough crazyness. In fact it is already leading to weird to debug errors.

johnbelamaric replied:
What errors? We should go ahead without it for this PR and then add another PR to do it, and discuss it there. To me, it seems we are responsible for making sure that the AXFR contains all the records we might return for the zone.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] [wip] support..." ]
>The above kind of works. I'm able to get transfers to work some of the times, other times it seems like the messages are being received by the secondary out of order which is causing the client to keep the connection open for >2s which is causing the readdeadline timeout to get reached. It _feels_ like the issue is with https://github.com/coredns/coredns/pull/1259/files#diff-347a95a2d36b7b336cb55cc146bd16d6R81 when we start writing the message back. I put some debugging in file/secondary.go and I can see the records being received out of order sometimes.  May need some help in tracking that down.

Do you have some info/details on this? For AXFR there is no out of order, only 
the first and last SOA matter.

/Miek

--
Miek Gieben

bradbeam commented:
Yeah, I added a simple print statement in https://github.com/coredns/coredns/blob/master/plugin/file/secondary.go#L34 to print out `env`

Works:
```
2017/12/18 05:34:28 [DEBUG] will we fail &{RR:[v803.puclf02-r01.cluster.local.	0	IN	A	10.178.245.7 v301.puclf01-r01.cluster.local.	0	IN	A	100.65.0.3 sp02-r01.cluster.local.	IN	A	10.30.0.17 v251.puclf01-r01.cluster.local.	0	IN	A	10.30.6.3 v300.puclf01-r01.cluster.local.	0	IN	A	100.65.0.1 v800.puclf01-r01.cluster.local.	0	IN	A	10.178.245.1 v255.puclf01-r01.cluster.local.	0	IN	A	10.30.6.11 puclf02-r01.cluster.local.	0	IN	A	10.30.0.11 v254.puclf01-r01.cluster.local.	0	IN	10.30.6.9 v253.puclf02-r01.cluster.local.	0	IN	A	10.30.6.7 v303.puclf02-r01.cluster.local.	0	IN	A	100.65.0.7 v256.puclf02-r01.cluster.local.	0	IN	A	10.30.6.13 sp01-r01.cluster.local.	0	IN	A	10.30.0.16 v252.puclf02-r01.cluster.local.	0	IN	A	10.30.6.5 v302.puclf02-r01.cluster.local.	0	IN	A	100.65.0.5 v801.puclf01-r01.cluster.local.	0	IN	A	10.178.245.3 v257.puclf02-r01.cluster.local.	0	IN	A	10.30.6.15 v802.puclf02-r01.cluster.local.	0	IN	A	10.178.245.5] Error:<nil>}
2017/12/18 05:34:28 [DEBUG] will we fail &{RR:[cluster.local.	0	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1513593277 7200 1800 86400 30] Error:<nil>}
2017/12/18 05:34:28 [INFO] Transferred: cluster.local. from 10.30.12.1:53
```

Failed
```
2017/12/18 05:34:24 [DEBUG] will we fail &{RR:[puc-vcsa01-witness.cluster.local.	0	IN	A	10.25.8.32 puc-vcsa01.cluster.local.	0	IN	A	10.25.3.11 v256.puclf02-r01.cluster.local.	0	IN	A	10.30.6.13 v250.puclf01-r01.cluster.local.	0	IN	A	10.30.6.1 v801.puclf01-r01.cluster.local.	0	IN	A	10.178.245.3 v252.puclf02-r01.cluster.local.	0	IN	A	10.30.6.5 v301.puclf01-r01.cluster.local.	0	IN	A	100.65.0.3 v303.puclf02-r01.cluster.local.	0	IN	A	100.65.0.7 v255.puclf01-r01.cluster.local.	0	IN	A	10.30.6.11 v254.puclf01-r01.cluster.local.	0	IN	A	10.30.6.9 v257.puclf02-r01.cluster.local.	0	IN	A	10.30.6.15 v802.puclf02-r01.cluster.local.	0	IN	A	10.178.245.5 v302.puclf02-r01.cluster.local.	0	IN	A	100.65.0.5 v253.puclf02-r01.cluster.local.	0	IN	A	10.30.6.7 v300.puclf01-r01.cluster.local.	0	IN	A	100.65.0.1 v251.puclf01-r01.cluster.local.	0	IN	A	10.30.6.3 v800.puclf01-r01.cluster.local.	0	IN	A	10.178.245.1 puclf01-r01.cluster.local.	0	IN	A	10.30.0.10 puclf02-r01.cluster.local.	0	IN	A	10.30.0.11] Error:<nil>}
2017/12/18 05:34:24 [DEBUG] will we fail &{RR:[cluster.local.	0	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1513593272 7200 1800 86400 30 sp01-r01.cluster.local.	0	IN	A	10.30.0.16 sp02-r01.cluster.local.	0	IN	A	10.30.0.17 v803.puclf02-r01.cluster.local.	0	IN	A	10.178.245.7] Error:<nil>}
2017/12/18 05:34:26 [DEBUG] will we fail &{RR:[] Error:read tcp 10.30.14.19:46276->10.30.12.1:53: i/o timeout}
2017/12/18 05:34:26 [ERROR] Failed to transfer `cluster.local.' from "10.30.12.1:53": read tcp 10.30.14.19:46276->10.30.12.1:53: i/o timeout
```

The out of order bits I'm referring to are specifically -- 
```
2017/12/18 05:34:24 [DEBUG] will we fail &{RR:[cluster.local.	0	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1513593272 7200 1800 86400 30 sp01-r01.cluster.local.	0	IN	A	10.30.0.16 sp02-r01.cluster.local.	0	IN	A	10.30.0.17 v803.puclf02-r01.cluster.local.	0	IN	A	10.178.245.7] Error:<nil>}
```
where we get a SOA record, and then two additional A records. 

For testing I also added in a debug statement here to print out `records` and the ordering looks correct there. https://github.com/coredns/coredns/pull/1259/files#diff-347a95a2d36b7b336cb55cc146bd16d6R77

miekg commented:
How does AXFR fail? How are you testing? With dig?

miekg commented:
OK, I've just pushed https://github.com/coredns/coredns/pull/1325 which shows a bit what I meant. Showing the code is prolly more efficient than keep on commenting on a PR.

Also while doing saw the Transfer method is kinda silly and this may have led you astray. I.e we just want to grab the records from k8s and pipe them through as an axfr. The msg channel seems a bit of an overkill...

Currently broken: the records returns (I just made things up), tests obviously. 

I'll try to see if this can mimic the errors you saw and hopefully debug that.

If we want to do that `fallthrough` thing we should do that as another PR whenever the xfr stuff is merged.

bradbeam commented:
The tests were between coredns instances.

```
.:53 {
          cache 30
          errors
          health
          loadbalance round_robin
          log stdout
          prometheus 
          proxy . 8.8.8.8 
}
cluster.local {
          kubernetes {
            transfer to 10.30.14.19
            fallthrough
          }
          file /etc/coredns/zones.d/cluster.local.db {
            transfer to 10.30.14.19
          }
          import /etc/coredns/conf.d/cluster.local.conf
}
```

The secondary:
```
.:53 {
  errors
  health
  loadbalance
  log
  prometheus 
}
cluster.local {
  debug
  
  secondary {
    transfer from 10.30.12.1
  }	
}
```

miekg commented:
hmm, interesting

bradbeam commented:
But dig does reproduce the issue as well

miekg commented:
OK.

Btw you dnsrecorder change makes sense as well, however I don't think (hope) we need locking here. I'll send a PR for changing the interface a bit and the caller.

bradbeam commented:
Added in an additional test for fallthrough to at least test for what I'm seeing. Going to save off this and trim this PR down to just kubernetes records support and follow up with a fallthrough bit.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] [wip] support..." ]
>Added in an additional test for fallthrough to at least test for what I'm seeing. Going to save off this and trim this PR down to just kubernetes records support and follow up with a fallthrough bit.

I won't merge this unless the fallthrough stuff is removed.

If we need this we should gather the extra records in the GetRecords() stage, 
not by abusing fallthrough like this.

bradbeam commented:
Understood, just wanted to finish this out train of thought. Working on removing the fallthrough bits.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] [wip] support..." ]
>Understood, just wanted to finish this out train of thought. Working on removing the fallthrough bits.

Ack. We can discuss "getting all the other records" in another PR. Some thought 
on that:

* Don't use `fallthrough` but have new code that grabs all the records we 
  want/need
* Failure mode: what if you can't resolv 1 external name, do you fail the axfr?
* Failure mode: what if an external name changes? Do you check for that? 

And lastly:

If you load this incomplete axfr into another CoreDNS with the secondary plugin 
and the same upstream configured: will it have all the data?

bradbeam commented:
I've removed the fallthrough bits, should be ready.

johnbelamaric commented:
@bradbeam looks like a race condition in Travis for the XFR tests - have you checked into it?

bradbeam commented:
Yup, got that. 
Next issue is it looks like my changes to dnstest with buffering the whole message and not the last line is causing failures in other plugins. Probably worth checking first which should be the correct behavior.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] [wip] support..." ]
>Yup, got that.
>Next issue is it looks like my changes to dnstest with buffering the whole message and not the last line is causing failures in other plugins. Probably worth checking first which should be the correct behavior.

please use dns.multireader

miekg commented on plugin/kubernetes/controller.go:
```diff
@@ -41,9 +42,14 @@ type dnsController interface {
 }
 
 type dnsControl struct {
+	modified int64 // modified tracks last modified
```
please add a comment that this needs to be first because we use `sync.LoadAtomatic` and therefor it needs to be on a 8 byte boundary.

miekg commented on plugin/kubernetes/controller.go:
```diff
@@ -41,9 +42,14 @@ type dnsController interface {
 	HasSynced() bool
 	Stop() error
+
+	// Modified returns last modified.
```
Think this is better phrases as: `Modified returns the timestamp of the most recent changes`

miekg commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,92 @@
+				for _, p := range svc.Spec.Ports {
+
+					s := msg.Service{Host: svc.Spec.ClusterIP, Port: int(p.Port), TTL: k.ttl}
+					s.Key = strings.Join([]string{zonePath, Svc, svc.Namespace, svc.Name}, "/")
```
I think we should return RRs here if possible instead of msg.Services, you also don't need `HostType` because you make that decision here.

This also means the transfer interface need to be changed, just like it did in my PR

bradbeam replied:
The one in #1325 ?

bradbeam commented on plugin/kubernetes/handler.go:
```diff
@@ -53,6 +57,37 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 		}
 		fallthrough
+	case dns.TypeAXFR:
+		records, err = plugin.AXFR(&k, zone, state, plugin.Options{})
```
As in create a new handler for `dns.TypeAXFR` and `dns.TypeIXFR` or move the `plugin.AXFR` out of `backend_lookup.go` ?

bradbeam replied:
Oh, I think I see what you mean based on that other PR. Following that.

bradbeam commented:
There's a little more cleanup left, but I think this should address most of the comments. The current test failures are due to the SRV record weight being set at 8080 versus the test cases using `25` and  `100`. Any concerns about changing these to 8080?

bradbeam commented:
Bummer, dont think my last comment made it through. Anyways, found out that it was a bad idea to just switch the weight to 100. There's two different code paths for generating SRV records, `msg.NewSRV`[1] and `plugin.SRV`[2]. Most tests utilize the dynamic weight/priority calculation of plugin.SRV, but since we're using msg.NewSRV, I needed to account for that. The path of least resistance was to copy the logic from plugin.SRV and add it to xfr.go. I'm guessing we'll need to find a better way to handle this. 

Anyways, I think I've got everything addressed and ci comes back green still. So I think we're in good shape. 

[1] https://github.com/coredns/coredns/blob/master/plugin/etcd/msg/service.go#L71-L76
[2] https://github.com/coredns/coredns/blob/master/plugin/backend_lookup.go#L152-L173

johnbelamaric commented:
@bradbeam is this ready for review again, all comments addressed?

bradbeam commented:
Yessir.

johnbelamaric commented on plugin/kubernetes/handler.go:
```diff
@@ -53,6 +53,39 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 		fallthrough
+	case dns.TypeAXFR, dns.TypeIXFR:
+		k.Transfer(ctx, state)
+		/*
```
Why is all this commented out code still here?

bradbeam replied:
Poor memory. I'll get it cleaned up

johnbelamaric commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,190 @@
+					// ref: https://github.com/coredns/coredns/pull/823
+					c <- s.NewSRV(msg.Domain(s.Key), 100)
+
+					s.Key = strings.Join(append(svcBase, strings.ToLower("_"+string(p.Protocol)), strings.ToLower("_"+string(p.Name))), "/")
```
Unnamed ports do not get an SRV record, according to the spec: https://github.com/kubernetes/dns/blob/master/docs/specification.md#232---srv-records

I think you are giving SRV records even for unnamed ports, unless I am reading this incorrectly?

bradbeam replied:
I think you're reading it correctly. I'll clean that up.

bradbeam commented:
@johnbelamaric updated

johnbelamaric commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,202 @@
+	rrs := make(chan dns.RR)
+	go k.transfer(rrs, state.Zone)
+
+	records := []dns.RR{}
```
Is everything from here down non-K8s specific? Maybe (after merge?) this could move to a utility function that other plugins could use, so they only have to implement a function that returns a channel?

miekg commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,202 @@
+	}
+
+	if len(records) == 0 {
+		return dns.RcodeServerFailure, nil
```
is this the right return code? (not super important, client will retry anyway).

bradbeam replied:
It's the same one that's implemented for file.xfr  https://github.com/coredns/coredns/blob/master/plugin/file/xfr.go#L30

miekg commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,202 @@
+					for _, addr := range eps.Addresses {
+						s := msg.Service{Host: addr.IP, TTL: k.ttl}
+						s.Key = strings.Join(svcBase, "/")
+						emitRecord(c, s)
```
why is the return value from emitRecord not used here , a few lines down it is used?
(warrants at least a comment)

miekg commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,202 @@
+
+// emitRecord generates a new A or AAAA record based on the msg.Service and writes it to
+// a channel. emitRecord returns the host name from the generated record.
+func emitRecord(c chan dns.RR, message msg.Service) string {
```
emitAddressRecord ? CNAME are sent using this for instance, nor SRV

miekg commented on plugin/kubernetes/xfr.go:
```diff
@@ -1,17 +1,202 @@
+
+// calcSrvWeight borrows the logic implemented in plugin.SRV for dynamically
+// calculating the srv weight and priority
+func calcSRVWeight(numservices int) uint16 {
```
prolly move this to a lib function as some point as well.

bradbeam commented:
Think I've got your comments addressed. I was a little unsure about the imports formatting, but gave it a shot. lmk if it's not quite right.

miekg commented:
let's not linger on that. I'm pressing buttons, stand back!
